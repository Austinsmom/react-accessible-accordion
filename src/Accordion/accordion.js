// @flow

import React, { Component } from 'react';
import { observable } from 'mobx';
import { Provider } from 'mobx-react';
import type { Node } from 'react';
import { isArraysEqualShallow } from '../utils';

type AccordionProps = {
    accordion: boolean,
    children: Node,
    activeItems: Array<string | number>,
    className: string,
    onChange: Function,
};

type AccordionState = {
    activeItems: Array<string | number>,
};

class Accordion extends Component<AccordionProps, AccordionState> {
    static defaultProps = {
        accordion: true,
        onChange: () => {},
        className: 'accordion',
        activeItems: [],
    };

    state = {
        activeItems: this.preExpandedItems(),
        accordion: true,
    };

    accordionStore = observable({
        activeItems: [],
        setActiveItems((foo) => this.activeItems = foo),
    })

    componentWillReceiveProps(nextProps: AccordionProps) {
        if (!isArraysEqualShallow(nextProps.activeItems, this.state.activeItems)) {
            let newActiveItems;
            if (nextProps.accordion) {
                newActiveItems = nextProps.activeItems.length
                    ? [nextProps.activeItems[0]]
                    : [];
            } else {
                newActiveItems = nextProps.activeItems.slice();
            }
            this.setState({
                activeItems: newActiveItems,
            });

            nextProps.onChange(nextProps.accordion ? newActiveItems[0] : newActiveItems);
        }
    }

    preExpandedItems() {
        let activeItems = [];
        React.Children.map(this.props.children, (item, index) => {
            if (item.props.expanded) {
                if (this.props.accordion) {
                    if (activeItems.length === 0) activeItems.push(item.props.customKey || index);
                } else {
                    activeItems.push(item.props.customKey || index);
                }
            }
        });
        if (activeItems.length === 0 && this.props.activeItems.length !== 0) {
            activeItems = this.props.accordion ? [this.props.activeItems[0]] : this.props.activeItems.slice();
        }
        return activeItems;
    }

    handleClick(key: number | string) {
        let activeItems = this.state.activeItems;
        if (this.props.accordion) {
            activeItems = activeItems[0] === key ? [] : [key];
        } else {
            activeItems = [...activeItems];
            const index = activeItems.indexOf(key);
            const isActive = index > -1;
            if (isActive) {
                // remove active state
                activeItems.splice(index, 1);
            } else {
                activeItems.push(key);
            }
        }
        this.setState({
            activeItems,
        });

        this.props.onChange(this.props.accordion ? activeItems[0] : activeItems);
    }

    render() {
        const { className, accordion } = this.props;
        return (
            <div role={accordion ? 'tablist' : null} className={className}>
                <Provider accordionStore={this.accordionStore}>
                    {this.props.children}
                </Provider>
            </div>
        );
    }
}

export default Accordion;
