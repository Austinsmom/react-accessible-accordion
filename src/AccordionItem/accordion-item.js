// @flow

import React, { Component } from 'react';
import type { Node } from 'react';
import { inject, observer, Provider } from 'mobx-react';
import consecutive from 'consecutive';
import classNames from 'classnames';
import type { Store } from '../accordionStore/accordionStore';

const nextUuid = consecutive();

type AccordionItemProps = {
    children: Node,
    className: string,
    hideBodyClassName: string,
    itemkey: string | number,
    accordionStore: Store,
    disabled: boolean,
    expanded: boolean,
};

class AccordionItem extends Component<AccordionItemProps, *> {
    static defaultProps = {
        className: 'accordion__item',
        hideBodyClassName: '',
        disabled: false,
        expanded: false,
    };

    customKey = this.props.itemkey || nextUuid();

    componentWillMount() {
        this.props.accordionStore.addItem({
            itemkey: this.customKey,
            itemUuid: nextUuid(),
            expanded: this.props.expanded,
            disabled: this.props.disabled,
        });
    }

    componentWillUnmount() {
        this.props.accordionStore.removeItem(this.customKey);
    }

    // This is here so that the user can dynamically set the 'expanded' state using the 'expanded' prop.
    componentWillReceiveProps({
        expanded,
        accordionStore,
    }: AccordionItemProps) {
        if (expanded !== this.props.expanded) {
            accordionStore.setExpanded(this.customKey, expanded);
        }
    }

    render() {
        const {
            className,
            hideBodyClassName,
            children,
            accordionStore,
        } = this.props;

        const currentItem = accordionStore.items.find(
            item => item.itemkey === this.customKey,
        );

        if (!currentItem) {
            return null;
        }
        const { expanded } = currentItem;

        return (
            <Provider itemkey={this.customKey}>
                <div
                    className={classNames(className, {
                        [hideBodyClassName]: !expanded && hideBodyClassName,
                    })}
                >
                    {children}
                </div>
            </Provider>
        );
    }
}

export default inject('accordionStore')(observer(AccordionItem));
