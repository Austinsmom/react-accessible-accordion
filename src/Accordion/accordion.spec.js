import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
// import ReactTestUtils from 'react-addons-test-utils';

import Accordion from './accordion';
import AccordionItem from '../AccordionItem/accordion-item';

jest.mock('../AccordionItem/accordion-item', () => 'div');

describe('Accordion', () => {
    it('renders correctly with min params', () => {
        const tree = renderer.create(
            <Accordion>
                <AccordionItem>Fake child</AccordionItem>
                <AccordionItem>Fake child</AccordionItem>
            </Accordion>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with accordion false', () => {
        const tree = renderer.create(
            <Accordion accordion={false}>
                <AccordionItem>Fake child</AccordionItem>
                <AccordionItem>Fake child</AccordionItem>
            </Accordion>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('disabled children', () => {
        const tree = renderer.create(
            <Accordion accordion={false}>
                <AccordionItem disabled={true}>Fake Child</AccordionItem>
                <AccordionItem>Fake Child</AccordionItem>
            </Accordion>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('handleClick function accordion true', () => {
        const wrapper = shallow(
            <Accordion>
                <AccordionItem>Fake Child</AccordionItem>
                <AccordionItem>Fake Child</AccordionItem>
            </Accordion>,
        );
        wrapper.instance().handleClick(1);

        expect(wrapper.instance().state.activeItems).toEqual([1]);

        wrapper.instance().handleClick(1);

        expect(wrapper.instance().state.activeItems).toEqual([]);
    });

    it('handleClick function accordion false', () => {
        const wrapper = shallow(
            <Accordion accordion={false}>
                <AccordionItem>Fake Child</AccordionItem>
                <AccordionItem>Fake Child</AccordionItem>
            </Accordion>,
        );
        wrapper.instance().handleClick(1);

        expect(wrapper.instance().state.activeItems).toEqual([1]);

        wrapper.instance().handleClick(1);

        expect(wrapper.instance().state.activeItems).toEqual([]);

        wrapper.instance().handleClick(0);
        wrapper.instance().handleClick(1);

        expect(wrapper.instance().state.activeItems).toEqual([0, 1]);
    });
});
