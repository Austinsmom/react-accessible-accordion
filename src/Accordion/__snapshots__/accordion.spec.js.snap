// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion disabled children 1`] = `
<div
  className="accordion"
>
  <div
    accordion={false}
    disabled={true}
    expanded={false}
    onClick={[Function]}
  >
    Fake Child
  </div>
  <div
    accordion={false}
    expanded={false}
    onClick={[Function]}
  >
    Fake Child
  </div>
</div>
`;

exports[`Accordion renders correctly with accordion false 1`] = `
<div
  className="accordion"
>
  <div
    accordion={false}
    expanded={false}
    onClick={[Function]}
  >
    Fake child
  </div>
  <div
    accordion={false}
    expanded={false}
    onClick={[Function]}
  >
    Fake child
  </div>
</div>
`;

exports[`Accordion renders correctly with min params 1`] = `
<div
  className="accordion"
>
  <div
    accordion={true}
    expanded={false}
    onClick={[Function]}
  >
    Fake child
  </div>
  <div
    accordion={true}
    expanded={false}
    onClick={[Function]}
  >
    Fake child
  </div>
</div>
`;
